test/com/google/enterprise/adaptor/DocIdSenderTest.java

```diff
@@ -347,74 +353,282 @@ public void getModifiedDocIds(DocIdPusher pusher) throws IOException, Interrupte
-
     config.setValue("adaptor.markAllDocsAsPublic", "true");
-    assertNull(docIdSender.pushGroupDefinitions(groups, false, null));
+    assertNull(docIdSender.pushGroupDefinitions(sampleGroups(), false, null));
 
     assertEquals(0, fileMaker.i);
     assertTrue(fileMaker.groupses.isEmpty());
     assertTrue(fileArchiver.failedFeeds.isEmpty());
   }
 
   @Test
+  @SuppressWarnings("unchecked")
+  public void testPushGroupsReplaceAllGroupsBeforeVersion74() throws Exception {
+    config.setValue("feed.maxUrls", "2");
+    assertNull(docIdSender.pushGroupDefinitions(sampleGroups(),
+        EVERYTHING_CASE_SENSITIVE, FULL, "foo", null));
+
+    assertEquals(2, fileMaker.i);
+    assertEquals(expectedResult(2, sampleGroups()), fileMaker.groupses);
+    assertEquals(ImmutableList.of("incremental", "incremental"),
+        fileSender.feedtypes);
+    assertEquals(ImmutableList.of("foo", "foo"), fileSender.groupsources);
+    assertEquals(ImmutableList.of("0", "1"), fileSender.xmlStrings);
+    assertEquals(ImmutableList.of("0", "1"), fileArchiver.feeds);
+    assertTrue(fileArchiver.failedFeeds.isEmpty());
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void testPushGroupsReplaceAllGroupsAtVersion740() throws Exception {
+    config.setValue("feed.maxUrls", "2");
+    config.setValue("gsa.version", "7.4.0-1");
+    assertNull(docIdSender.pushGroupDefinitions(sampleGroups(),
+        EVERYTHING_CASE_SENSITIVE, FULL, "foo", null));
+
+    assertEquals(3, fileMaker.i);
+    assertEquals(expectedResult(2, sampleGroups(), emptyGroups()),
+        fileMaker.groupses);
+    assertEquals(ImmutableList.of("incremental", "incremental", "full"),
+        fileSender.feedtypes);
+    assertEquals(ImmutableList.of("foo-FULL1", "foo-FULL1", "foo"),
+        fileSender.groupsources);
+    assertEquals(ImmutableList.of("0", "1", "2"), fileSender.xmlStrings);
+    assertEquals(ImmutableList.of("0", "1", "2"), fileArchiver.feeds);
+    assertTrue(fileArchiver.failedFeeds.isEmpty());
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void testPushGroupsReplaceAllGroupsSingleFeed() throws Exception {
+    config.setValue("gsa.version", "7.4.0-1");
+    assertNull(docIdSender.pushGroupDefinitions(sampleGroups(),
+        EVERYTHING_CASE_SENSITIVE, FULL, "foo", null));
+
+    assertEquals(2, fileMaker.i);
+    assertEquals(
+        expectedResult(Integer.MAX_VALUE, sampleGroups(), emptyGroups()),
+        fileMaker.groupses);
+    assertEquals(ImmutableList.of("incremental", "full"), fileSender.feedtypes);
+    assertEquals(ImmutableList.of("foo-FULL1", "foo"), fileSender.groupsources);
+    assertEquals(ImmutableList.of("0", "1"), fileSender.xmlStrings);
+    assertEquals(ImmutableList.of("0", "1"), fileArchiver.feeds);
+    assertTrue(fileArchiver.failedFeeds.isEmpty());
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void testPushGroupsReplaceAllGroupsAlternateBuffer() throws Exception {
```

Comment:
Added new tests:
1) First 2/3 of testPushGroupsIncrementalThenFullThenFull
2) testPushGroupsIncrementalThenFullThenFull
3) testPushGroupsFullThenIncrementalThenIncremental
4) First 3/4 of testPushGroupsFullThenIncrementalThenFullThenIncremental
5) testPushGroupsFullThenIncrementalThenFullThenIncremental

I also added a test for feeding from multiple group sources.
These additional tests are exercising the previousGroupSource()
and alternateGroupSource() methods for selecting which feed
datasource name to use.